Updates to modifiers proposal

ChickenProp opened the pull request:
The [proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0370-modifiers.rst) has been [accepted](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0111-linear-types.rst); the following discussion is mostly of historic interest

--------------------------------------------

[Rendered](https://github.com/ChickenProp/ghc-proposals/blob/modifiers-updates/proposals/0370-modifiers.rst)

I've been [working on](https://gitlab.haskell.org/ghc/ghc/-/issues/22624) implementing the modifiers proposal #370. The current proposal seemed to have some mistakes and omissions. Here are my suggestions for amending it.

Summary of the diff (as of October 18):

* ';' is now allowed in between modifiers attached to decls, and required before the decl. So

      %One   %Two ; data F -- previously allowed, still allowed
      %One   %Two   data F -- previously allowed, now forbidden
      %One ; %Two ; data F -- previously forbidden, now allowed
      %One ; %Two   data F -- previously forbidden, still forbidden

  Initially, I made the one before the decl required for parsing reasons. I couldn't figure out how to have all of: optional pre-decl semicolon; modifiers attached to expressions; expressions can appear at the top level. (I was getting shift/reduce or reduce/reduce conflicts in the parser.) Since there are no longer modifiers attached to expressions, it could be optional again. But in case we want modifiers on expressions in future, we might want to keep our options open.

* No modifiers attached to types or expressions.

* `-Wunrecognized-modifiers`, not `-Wunknown-modifiers`. Specify that it's enabled by default.

* `-XLinearTypes -XNoModifiers` introduces backwards compatible behavior.

* Add non-normative discussion of RN versus TC modifiers.

* In the linear-types proposal, update the examples of modifiers attached to lambda-bound variables. I think the examples now match the BNF given. Also, the line "the `lpat` must be a bare variable" was contradicted by syntax elsewhere in the proposal, so remove that.

ChickenProp committed:
Update modifiers proposal.

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -54,37 +54,40 @@ Proposed Change Specification
+     prefix%       ::= '%'    -- only in prefix position
+     modifier      ::= prefix% atype
+     modifiers     ::= {modifier}
+     declModifiers ::= [ modifier [ ';' ] [ declModifiers ] ]
```
The proposal as-is allows

```
%One
  %Two
data F

%One
  %Two
  data F
```

but forbids

```
%One
%Two
data F

%One
%Two
  data F
```

I propose that these be allowed too. The new `declModifiers` allows semis in between modifiers that precede declarations.

ChickenProp replied:
It turns out that

```
%One
  %Two
  data F

%One
%Two
  data F
```

are difficult to parse. The problem is that you can put an expr at the top level, as a naked TH splice, and that leads to parse conflicts. I don't have a solid grasp of what is and isn't possible, so maybe it can be allowed in future. But the current implementation just allows

```
%One
  %Two
data F

%One
%Two
data F
```

I'm not sure if I should update the proposal to reflect that, or leave it as-is aspirationally.

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -54,37 +54,40 @@ Proposed Change Specification
+     topdecl ::= declModifiers 'type' simpletype '=' type
+             |   declModifiers 'data' [context '=>'] simpletype ['=' constrs] [deriving]
+             |   declModifiers 'newtype' [context '=>'] simpletype = newconstr [deriving]
+             |   declModifiers 'type' 'data' ...
```
There's a bunch of places the proposal doesn't currently allow modifiers where we might want them, and I don't think we need to add them everywhere. But `type data` seemed like a close enough fit with type/data/newtype that it was worth adding now.

Presumably I need to be more precise about what the `...` is. But I confess I'm not sure where I'd look up the currently accepted BNF here. (Maybe it's just the same as for regular `data`?) Can someone give me a pointer?

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -54,37 +54,40 @@ Proposed Change Specification
+             |   declModifiers 'instance' [scontext '=>'] qtycls inst ['where' idecls]
+             |   declModifiers 'default' '(' type1 ',' ... ',' typen ')'
+             |   declModifiers 'foreign' fdecl
+             |   declModifiers ';' decl
```
I guess this isn't technically the right syntax, because it allows two semis between the modifier and the declaration:

    %One ;; foo :: Int

I can fix this if `declModifiers` looks like it's going to be accepted.

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+* The ``%1`` modifier is handled as a special case. It's renamed (and so
+  typechecked) the same as ``%One``, even if it appears somewhere that linear
+  modifiers aren't expected. If a user does want the modifier ``1 :: Nat``, they
+  can write it as ``%01``. (Requires ``-XDataKinds``.)
```
The previous proposal was unclear about %1, saying both:

> there will be precisely one modifier: the `%1` or `%Many` of linear types

> The `%1` will remain a single lexeme and does not participate with this proposal.

I wasn't sure what it meant for it to not participate with this proposal, given that the user might write it in any place where a modifier is expected. @int-index [suggested](https://gitlab.haskell.org/ghc/ghc/-/issues/22624#note_582305) to have `%1` be special syntax for `%One` if and only if `-XLinearTypes` is enabled.

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+
+``-XLinearTypes`` implies ``-XModifiers``. But the latter can be explicitly
+disabled with ``-XLinearTypes -XNoModifiers``. This introduces backwards
+compatible behavior:
```
My *intent* is for this to be backwards compatible, so that users could keep existing programs working by adding `-XNoModifiers`. I'm not super confident I succeeded.

I don't have a strong opinion on whether we want this temporarily, permanently, or not at all. I lean "not permanently", but it's nice that it gives an easier migration route.

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+This behavior may be deprecated in future.
+
+Renaming and typechecking
+-------------------------
```
I think this is what @goldfirere [suggested](https://gitlab.haskell.org/ghc/ghc/-/issues/22624#note_469566) to handle this problem.

goldfirere replied:
Certain modifiers (that is, modifiers of certain kinds) apply at renamer-time. We can exhaustively list them. (I don't have the full list in my head at the moment, but GHC will always have a fixed set of kinds that can describe modifiers, and some subset of that list affect the renamer.) So that's the specification. If it turns out that the type-checker encounters a modifier whose kind works out to be one of the RN-modifiers (but the modifier wasn't spotted by the renamer, because it required type inference and/or type family/synonym reduction), error.

An alternative approach is to have TC-modifiers use `%`s and RN-modifiers use `&`s, say. Maybe that's a better user experience? I tend to think not, especially because a type-checker-dependent attempt to use an RN-modifier will produce a very clear error message.

sgraf812 replied:
Personally, I find it deeply questionable to allow anything other than normal forms in modifier position, but it's not a hill I'm willing to die on.

We could say that uses of RN modifiers need to be enums. These are very simple to type-check; it's just name analysis.

googleson78 commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
-    know that a modifier is unknown, it still must be parsed, renamed, and type-checked.)
+13. Modifiers of unknown kind produce an error.
+
+14. Modifiers of known kind but with an unknown meaning produce a warning,
+    controlled by ``-Wunknown-modifiers``. They are otherwise ignored. (However,
+    in order to know that a modifier is unknown, it still must be parsed,
+    renamed, and type-checked.)
```
This is not something that your changes here introduce, but how is "unknown meaning" defined?

Does this allow for a potential future where user provided modifiers exist?

ChickenProp replied:
I guess something like, "the default behavior of modifiers is to emit this warning, but if GHC specifically recognizes a modifier in some context it overrides that behavior".

Not sure about user provided modifiers. There's speculation about handling them via plugin, and then maybe it would be possible for that plugin to mark them as "I've done something with this modifier, no need to warn about it"? But I don't know enough about plugins to know if that's possible.

googleson78 replied:
In my idealised world, which I believe I previously shared, if modifiers are available in the metadata for `GHC.Generics.Generic`, you would not need to do anything with plugins - you would simply be able to access them via the metadata in `Generic`, same as with any other metadata about types in `Generic`. 

This is more of an "open world" sort of deal, because now, downstream consumers of your library can see if you've put some modifiers on your type and then potentially consume them in some way, which, now that I've said that, kind of implies to me that this mode of operation (consider unknown-to-ghc modifiers to be "bad") conflicts with what I've been imagining.

googleson78 commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+
+Renaming and typechecking
+-------------------------
+Initially, the only recognized modifiers will be the multiplicity modfifiers of
```
```suggestion
Initially, the only recognized modifiers will be the multiplicity modifiers of
```

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+A more complicated solution involves invoking the typechecker during renaming.
+To do this we need a "stage restriction": a modifier obeys the stage restriction
+if every identifier and every type family instance in the modifier is imported,
+not defined in the current module.
```
I'm not sure if this stage restriction is standard when dealing with TH or specific to modifiers. IIRC you can do `$(pure ())` at the top level to break a module into two compilation units. Would we expect that to work here, too?

Can we tell whether a type family instance comes from this module or not? Or would it be "we try to typecheck, and if it needs a type family instance that's not defined outside this module, that fails"?

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -167,8 +268,7 @@ Further examples:
 
-* Lambda expressions: ``\ x %Many -> ...`` or ``\ x %One -> ...``. This would be parsed but rejected, because
-  the new syntax applies only for lambda that bind a single, top-level variable: ``\ x y %One -> ...``.
+* Lambda expressions: ``\ %Many x -> ...``, ``\ %One x %Many y -> ...``.
```
Updated to match the syntax (not yet implemented) from the linear types proposal.

I'm not sure if this is accepted, or if it would need to be `\ (%One x) (%Many y) ->`.

aspiwack replied:
I think it's fine (better, even) to require parentheses. It's more consistent with type annotation. Clearer to parse. It's the right way. Feel free to modify the linear types proposal accordingly.

ChickenProp replied:
Hm, so on the question of what the specification *should* be: I don't feel strongly about parens being required when there's more than one argument. But when there's just one, I do feel like I prefer

    \ %One x ->

being allowed, rather than requiring

    \ (%One x) ->

and if the first one is allowed, then I'm not sure if it's feasible to require parens when there are multiple arguments. (That's actual not sure, not intended as "I don't think this is feasible".)

aspiwack replied:
The thing is, we already have:

```
> :t \ x::Int -> Int

<interactive>:1:4: error: [GHC-58481] parse error on input ‘::’
```

So it's worth being consistent. And it avoids a special case. Frankly this is a reasonable compromise. Even if `\ %1 x ->` does look better, I agree.

ChickenProp replied:
Yeah, that makes sense. And if we do want to be more permissive later we can do that. (I definitely don't want a special case so that `\ %1 x ->` is allowed but `\ %1 x %1 y ->` isn't, but I don't have strong feelings about allowing `\ %1 x %1 y ->`.)

Ah, but, do you know how I'd amend the BNF in the linear types proposal? Currently the relevant section is:

> Lambda-bound variables can also be annotated with a multiplicity:
>
> ```haskell
> \ (%'One x) :: A -> x
> ```
>
> is the identity function at type `A ⊸ A`. A binder can be annotated with a multiplicity without a type like this
>
> ```haskell
> \ %'One x -> x
> ```
>
> This modifies the syntax entry for lambda expressions as follows
>
>     lpat -> [PREFIX_PERCENT btype] lpat
>
> where the btype after the % must be of kind Multiplicity (see below). The lpat must be a bare variable, or an error occurs.

And I don't know what the current BNF is, to know what it should be changed to.

Also, I think the first example there was probably a mistake? Since as you note, `\ x :: A ->` is currently forbidden, so it would be weird if `\ (%1 x) :: A ->` was allowed. So I guess I'll change these examples to

```haskell
\ (%1 x :: A) -> x
\ (%1 x) -> x
```

(I assume we don't want to require `\ ((%1 x) :: A) ->`.)

aspiwack replied:
Yeah, I think `(%1 x :: A)` is fine. It really doesn't look problematic from a parsing point of view. If it is, we can revisit.

For the BNF, imagine that the linear types proposal don't exist. So the modifiers proposal is retroactively a dependence of the linear types implementation.

ChickenProp replied:
So, the problem I have is that I don't know what the BNF for the currently accepted syntax is. I found https://www.haskell.org/onlinereport/haskell2010/haskellch10.html but that doesn't allow `::` in patterns, that's a GHC extension. The extension is documented at https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/scoped_type_variables.html#pattern-type-sigs but not with BNF. (And it's not clear exactly which patterns allow signatures without being parenthesized and which don't; 

I could try to reverse engineer the BNF from the description, or from the parser code, but if there's an existing BNF written down somewhere, I'd be unlikely to match it exactly. Or I suppose I could try to look up where ScopedTypeVariables was first proposed and see if that has BNF?

But... I kinda feel like most people writing proposals probably don't do that kind of thing, and either know somewhere to look up the currently supported BNF, or just kinda guess?

If I'm guessing, then... well, somewhat embarrassingly, I now see that the current modifiers proposal has `lpat ::= modifiers lpat | ...`, and that feels like it might already be exactly what we want here. And I think it's the same BNF change as in the linear types proposal, just with different BNF syntax. Now that I know roughly what `lpat` is (it turns out it's not "the pattern inside a lambda", which is an `apat`), I see that yeah, the BNF given in the current LT proposal doesn't support the syntax examples in it.

All of which is to say, I currently think the changes I need to make are

* Update the examples in the linear types and modifiers proposals, showing that the syntax inside lambdas is `\ (%1 a) ->` and `\ (%1 a :: A) ->`.
* Remove the bit about "The `lpat` must be a bare variable" in the LT proposal, since that's contradicted by e.g. support for `let %1 !(x, y) = ...`. (It's kind of confusing that the BNF for patterns-in-general is described in a section talking specifically about lamba variables, but I think I'm not going to try fixing that.)

ChickenProp replied:
Done.

aspiwack replied:
> It's kind of confusing that the BNF for patterns-in-general is described in a section talking specifically about lamba variables, but I think I'm not going to try fixing that

No need, indeed.

> So, the problem I have is that I don't know what the BNF for the currently accepted syntax is.

What we do, in practice, is provide a diff with respect, either to the last Haskell report when it's still close enough to the current parser (but as the report is getting older, it becomes less and less relevant). Or simply open the parser proper and define a diff to that (in stylised BNF form).

Thanks for doing it.

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -184,13 +284,15 @@ Effect and Interactions
+* Proposals `#390`_ and `#512`_ also anticipate using modifier syntax.
+
+.. _`#390`: https://github.com/ghc-proposals/ghc-proposals/pull/390
+.. _`#512`: https://github.com/ghc-proposals/ghc-proposals/pull/512
```
Should I link #232 here as well? The author [thinks](https://github.com/ghc-proposals/ghc-proposals/pull/232#issuecomment-941854237) it's mostly a good fit for modifier syntax, and I agree. But it felt weird to be like "here's this other proposal that as-accepted has no relation to this one, but maybe could be changed".

goldfirere replied:
I think showing more ways in which modifiers can be helpful only makes the proposal stronger.

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -292,8 +394,12 @@ Unresolved Questions
   proposed concretely here and would be subject to a future proposal.
 
-.. _`#242`: https://github.com/ghc-proposals/ghc-proposals/pull/242
+* How does this interact with Template Haskell?
```
I don't know TH very well. I think the options might be something like:

1. TH types don't change until a modifier that would need them to change is introduced (so not with this proposal). Inside a splice, I think that means

   * At most one modifier is allowed on an arrow
   * Modifiers elsewhere can be either ignored or can throw an error

2. TH types change to get a list of modifiers, any place where modifiers are allowed.

ChickenProp committed:
Fix typo in proposals/0370-modifiers.rst

Co-authored-by: Georgi Lyubenov <godzbanebane@gmail.com>

goldfirere commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+
+.. _`proposal 512`: https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0512-nofieldselectors-per-datatype.md
+
+This poses a problem. The following is accepted::
```
Won't `Many` and `NoFieldSelectors` have different kinds?

ChickenProp replied:
Oh, right. I wasn't thinking about that, I just wanted to keep the example concise. I'll change it so that it's also correct.

goldfirere commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+It might resolve to a different modifier if there are overlapping instances
+defined in this module.
+
+This solution would probably not be implemented in the initial release of
```
I think I would take this little section and move it to Alternatives. That is, let's just say, for now, that RN modifiers need not to involve type synonyms or type families. This little algorithm might allow us to improve things in the future, but I think we should wait until there is enough real pain around the status quo before doing so.

goldfirere reviewed (commented):
Thanks for putting the time into these revisions! I like the direction!

int-index commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+With ``-XNoLinearTypes``, the ``%1`` modifier is not special. It refers to the
+type ``1 :: Nat`` and requires ``-XDataKinds``.
+
+``-XLinearTypes`` implies ``-XModifiers``. But the latter can be explicitly
```
> Though I imagine others will disagree, I favor 3.

Indeed, I would oppose that. `Modifiers` would be a new extension, and we should mark it "Experimental", and we can't impose experimental extensions on users by default. If, on the other hand, we mark it "Stable", that would create obstacles to making changes to this new extension.

Personally I think there are plenty of use cases that modifiers could cover (e.g. replace various instance pragmas) that we should specify and implement before `-XModifiers` fossilizes and gets enabled by default.

goldfirere replied:
My thought is that we should feel free to add more functionality to `-XModifiers` even if it is enabled by default -- as long as all the changes are backward compatible. But actually we might decide we made a mistake somewhere and *remove* functionality from `-XModifiers`, which suggests it should remain off by default for now. So I think I favor (1), of my three options above: don't have `-XLinearTypes` and `-XModifiers` interact directly, but allow the `%` syntax with either `-XLinearTypes` *or* `-XModifiers` (though with only the latter, it would cause an error later).

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+  modifier still has no meaning.
+
+* The ``%1`` modifier is handled as a special case. It's renamed (and so
+  typechecked) the same as ``%One``, even if it appears somewhere that linear
```
From base. I'll disambiguate this.

ChickenProp commented on proposals/0370-modifiers.rst:
````diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+  modifiers aren't expected. If a user does want the modifier ``1 :: Nat``, they
+  can write it as ``%01``. (Requires ``-XDataKinds``.)
+
+* The linear arrow ``a ⊸ b`` has the same meaning as ```a %1 -> b``. Other
````
Since `%1` always refers to the `One` from base, this does too.

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+* Only ``Multiplicity`` modifiers are permitted, and only in the places they're
+  recognized. Any use of a modifier is an error.
+
+* The kind of a modifier is determined by checking for ``Multiplicity``, not
```
Ah, can you elaborate on how it's not? (I see it should have been "any other use" on 167, but I guess that's not what you mean.)

I'm also not sure I see the connection. I think that if "linear types without modifiers" isn't backwards compatible, that reduces the motivation to have special rules for that case. (As opposed to, say, "without modifiers, linear types has no effect simply because you can't access any of the behavior changes it implies".) But I don't see why it reduces the motivation to have `-XLinearTypes` imply `-XModifiers` by default. If anything I think it increases that motivation.

goldfirere replied:
I can write `Int %m -> Int` today, and GHC will infer `m :: Multiplicity`. That would not be true with the more general `-XModifiers` approach.

ChickenProp replied:
Right, it might not have been clear: this section is specifically intended as "if the user wants backwards compatibility, here's how they can get it". So (according to the current state of this PR):

* With linear types and modifiers both enabled, we use synthesis-not-checking, and `Int %m -> Int` is an error.
* With linear types enabled and modifiers disabled, we go back to checking-not-synthesis, and `Int %m -> Int` is accepted. My intent is that this case is backwards compatible. I lean towards not keeping this around permanently, if we have it at all.
* To unlock that case, the user needs to write `-XLinearTypes -XNoModifiers`, which is admittedly a bit arcane.

simonpj reviewed (commented):
I drafted these comments many days ago, but failed to commit. Sorry!

sjoerdvisscher commented:
I am [implementing](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13525) multiplicity annotations on record fields at the moment, and I feel there needs to be a change there, maybe we can combine that change with this PR? Let me know if a separate proposal is preferred!

The current syntax is

```
data Record = Rec { x %'Many :: Int, y %1 :: Char }
```

But I would propose to change this to:

```
data Record = Rec { %'Many x :: Int, %1 y :: Char }
```

For two reasons:

1. Consistency, most of the time modifiers are placed in front of the name, f.e.:

```
let %'Many x :: Int = 1; %1 y :: Char = 'c' in ...
```

2. Preventing confusion on what the annotation means. This may the personal, but the original syntax suggested to me that the type of the selector changes: `x :: Record %'Many -> Int` and `y :: Record %1 -> Char`. But that's not what is happening, it is the constructor that changes: `Rec :: Int %Many -> Char %1 -> Record`. (The multiplicity of the selector can be derived: it can be linear if all the other fields are unrestricted, i.e. they are allowed to be thrown away, otherwise it is Many.)

Possibly there might be a need in the future for two types of modifiers on record fields, one that applies to the constructor and one that applies to the selector. Then maybe both places should be allowed?

simonpj commented:
> So there are at least multiplicity and matchability that we want to specify.

OK good point.    Maybe this:

* All modifiers must parse and rename as types.  Notably, they must mention only types that are in scope.
* After that, the behaviour is determined per-modifier.  Specifically:
  * `%NoFieldSelectors` is valid only in places X and Y; and must resolve to the type `GHC.Type.NoFieldSelectors`; no type famlilies,, no synonyms, nothing
  * `%mult` where `mult :: Multiplicity` is valid only in places P and Q (notably function arrows). The type is a first-class type, and can use type synonyms, type families etc.

So rather that attempting a general "RN modifier" story, just lay out the story per modifier.

goldfirere commented:
> So rather that attempting a general "RN modifier" story, just lay out the story per modifier.

Sure.... but all modifiers that affect renaming behavior will have the same such story, so it might be easier to name the story "the RN-modifier story" and re-use it. That said, we could start by having per-modifier explanations and wait for this convergence to happen instead of predicting in advance (and possibly getting that wrong).

JakobBruenker commented on proposals/0370-modifiers.rst:
```diff
@@ -54,37 +54,40 @@ Proposed Change Specification
-     type     ::= btype [ {modifier} -> type ]
-     prefix%  ::= '%'    -- only in prefix position
-     modifier ::= prefix% atype
+     type          ::= btype [ {modifier} -> type ]
```
Since `modifiers` is defined anyway, one could make this (very minor) simplification
```suggestion
     type          ::= btype [ modifiers -> type ]
```

ChickenProp committed:
Update proposals/0370-modifiers.rst

Co-authored-by: Jakob Brünker <10101851+JakobBruenker@users.noreply.github.com>

ChickenProp committed:
Small updates.

ChickenProp commented:
(Made some small changes, there are larger ones still to be come.)

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+  modifiers are accepted: ``a %Matchable ⊸ b`` has the same meaning as
+  ``a %Matchable %1 -> b``.
+
+With ``-XNoLinearTypes``, the ``%1`` modifier is not special. It refers to the
```
I've written it so that the `-Wunknown-modifiers` warning hints to enable linear types, rather than having a separate warning for this - is that good?

ChickenProp replied:
Oh, there's a wart here... if the user has `-XModifiers -XNoLinearTypes -XNoDataKinds`, then there's no "unknown modifier" warning because there's an "illegal type" error.

Maybe it's possible to add this hint to the "illegal type" error if that gets thrown, and to the "unknown modifier" warning if not?

Or maybe this is fine for a first pass? Then the user will get hinted to turn on `DataKinds`, and after they do that they'll be hinted to turn on `LinearTypes`, which isn't ideal but maybe it's acceptable.

ChickenProp commented:
@sjoerdvisscher

I don't feel strongly about that change. I like that it feels more consistent with

```
%x; f :: Int
data F where %x F :: F
let %x f = ()
```

But I'm concerned that in

```
data F = F { %a x, y :: Int }
```

it might be less clear that `%a` modifies both `x` and `y`.

sjoerdvisscher commented:
> But I'm concerned that in
> 
> ```
> data F = F { %a x, y :: Int }
> ```
> 
> it might be less clear that `%a` modifies both `x` and `y`.

Isn't this the same with regular type signatures?

jvanbruegge commented:
Not really. In a record all fields need a type signature, so it makes sense to assume that `:: Int` applies to both. However, modifiers are optional, so I would assume that it only applies to `x` above.

googleson78 commented:
`%mod name :: ty` also raises the question of precedence to someone who is unfamiliar with the syntax i.e. is it `(%mod name) :: ty` or `%mod (name :: ty)`

goldfirere commented:
@simonpj 

> Now, all modifiers share one characteristic: they must be parsed and renamed as types, including scoping. I'm with you there. But did you mean anything else?

Yes. All modifiers that affect renaming behavior will be unable to cope with type family/synonym reduction.

simonpj commented:
> Yes. All modifiers that affect renaming behavior will be unable to cope with type family/synonym reduction.

But we have to provide a spec for `%NoFieldSelectors`  *anyway*.  In that spec we just say "you have to say `%NoFieldSelectors`; you cannot use a type synonym or type family".  

So maybe it just comes down to this: 
* For some modifiers (call them TC modifiers) you can expect the program to behave the same if you replace a modifier by an equal type (in the sense of "nominally equal in the type system").  Example: linear modifiers on arrow `a %One -> b`
* For others (RN modifiers) you cannot; you must use precisely the type constructor in the spec for that modifier.  Example: `%NoFieldSelectors`.

To tell which of the two cases hold, you must look at the spec of the individual modifier.

Now it makes more sense.  Can it be framed this way and made fully part of the Change Specification?

ChickenProp committed:
Updates.

ChickenProp committed:
Editing RST.

ChickenProp commented:
I currently have it as "future proposals will have to specify how their modifiers behave, but we expect it'll look like...". I've left the more complicated thing there rather than moving it to alternatives, since the whole section is now non-normative and I felt like it flowed better keeping it all in one place.

Not currently planning to make more changes without further discussion.

ChickenProp commented:
> Isn't this the same with regular type signatures?

I guess the required semicolon helps here:

```
%declMod
a, b :: Int
a = 1
b = 2

-- versus
data Rec = Rec { %fieldMod a, b :: Int }
```


...but also, modifiers on decls are scoped more broadly than modifiers on record fields (`declMod` could presumably modify the `a = 1` and `b = 2`, where there's no equivalent of those in record fields), so maybe it makes sense that they're in different places? But then we have to choose between

```
-- Consistent with record fields
data A where
  A, B %foo :: A

-- Consistent with regular data declarations
data A where
  %foo A, B :: A
```

I kind of wonder if we should just not have modifiers on decls until someone wants one, and then we can figure out where they should go. But we'd still need to choose what to do with record fields and gadt constructors now.

sjoerdvisscher commented:
@ChickenProp You almost had me convinced, but your last example makes me wonder: Should we really introduce a rather big inconsistency in syntax just for a marginally better readability of a little used feature?

ChickenProp commented:
Yeah, I don't know.

I think my current position is: I don't personally have a strong feeling about where these should be, and they've already been accepted in the `foo %a :: Int` position. I won't be sad if that changes, and I'm happy for you to push for that change either here or in a separate proposal, but I'm not currently inclined to push for it myself.

ChickenProp committed:
Require parens around modified lambda vars.

ChickenProp committed:
Ask about scoping.

ChickenProp commented on proposals/0370-modifiers.rst:
```diff
@@ -292,8 +404,21 @@ Unresolved Questions
+
+* What warning groups imply ``-Wunknown-modifiers``?
+
+* What scope are modifiers looked up in? For example::
+
+    %a data F a
+    %G data G
+
+  If these are accepted, they'd be meaningless under the current proposal. But
+  should they be rejected, or accepted with an unknow-modifiers warning, or
+  what?
```
My current implementation allows using a modifier on its own definition:

```haskell
%CT1 data CT1

%CT2Con data CT2 = CT2Con

%CT2Con' data CT2' where
  CT2Con' :: CT2'

%CT3 class CT3 a
```

And has inconsistent behavior for free variables:

```haskell
-- accepted
%a data FV1 a

-- accepted
data FV2 a = %a FV2

-- rejected
data FV2' a where
  %a FV2' :: FV2'

-- accepted
data FV3 a = FV3 { fv3 %a :: () }

-- rejected
data FV3' a where
  FV3' :: { fv3' %a :: () } -> FV3

-- accepted, but seems especially sketchy
data FV4 = %a forall a . FV4 a

-- accepted
data FV4' where
  %a FV4' :: a -> FV4'

-- rejected
data FV5 = FV5 { fv5 %a :: forall a . a }

-- rejected
%a class FV6 a

-- rejected
%a instance B a
```

The easiest for me is to just continue not to think about this, and if in future it ever matters, we can decide then how it should behave. Downside is, at that point we might decide that programs currently accepted should be rejected. Maybe that's fine on grounds of "this is kind of experimental and no one has any reason to be writing those programs"?

ChickenProp committed:
Fix typo.

ChickenProp committed:
Some more examples.

ChickenProp committed:
Clarify that polymorphic kinds are also an error.

ChickenProp commented:
I've added some more examples, and clarified that e.g. `%Nothing` is an error because `Nothing` is poly-kinded.

ChickenProp committed:
Fix typo.

ChickenProp committed:
Use flag -Wunrecognised-modifiers instead of -Wunknown-modifiers.

adamgundry commented:
Thanks very much for your work implementing this @ChickenProp, and sorry for the delay. I've asked @Tritlo to shepherd this.

ChickenProp commented:
> Is it possible to summarise, in a list, what substantive (no just typos or formatting) changes you are proposing?

Yes, good idea. I'll have to go back through to double check, but I'll work on this. I'll also try to summarize where the current implementation falls short. Tritlo's list is probably pretty close, although

> Multiple modifiers are now formally allowed in one position (types, binders, declarations).

I think both the accepted version of the proposal and my amendment allow multiple modifiers everywhere. (I reworded from `{modifier}` to `modifiers` in the BNF, so it might have looked like allowing multiplicity.)

> Now that stacking multiple modifiers is allowed, what prevents users from writing %One %Many? The grammar allows it, but semantically it doesn’t make sense.

It's an error at typechecking time:

    ghci> let f :: a %1 %1 -> a; f = id
    <interactive>:8:10: error: [GHC-65505]
        • Too many Multiplicity modifiers
        • In the type signature: f :: a %1 %1 -> a

> Clarifying warnings: is `-Wunrecognized-modifiers` enabled by default? If I have `-XModifiers` and accidentally misspelled `%Manu`, is it just silently ignored? Will I see a warning or error without additional flags?

Not yet specified. I think it should be enabled by default, but I forgot about that until just now.

> Could you clarify exactly what layouts are allowed?

I think this is to do with parsing-in-general rather than this proposal, though I can also write it explicitly. If I understand parsing right, all three of the things you wrote would be accepted, and are in the current implementation. (But the current implemantation forbids

```haskell
%A;
data D
```

which I think should be allowed, and should be an easy fix.)

> I worry that special casing `%1` here is confusing, it’s strange that `%1` and `%2` will mean completely different things. I suggest that this should only be allowed for `-XLinearTypes -XNoModifiers` That’s not really for this proposal, but could you spell this out explicitly in the proposal, just to make it clear?

To clarify, you mean "spell out explicitly that `%1` and `%2` will mean completely different things"?

It is strange, but my sense is that having to switch from `%1` to `%One` would be significantly annoying for people who want to use linear types.

ChickenProp commented:
Yes, I'll add these.

> or I think this would also be allowed (true?)
>
>     instance %Overlappable C a => C [a] where ... [let's call this *]

Good question! Mostly false but it's complicated, and turns out there's at least one bug.

True in that [*] is syntactically valid, by virtue of "modifiers can be attached to types". But right now it parses the same as

    instance (%Overlappable C) a => C [a] where ...

...which is probably not what anyone wants or would expect. I'm not sure if I've implemented the proposal correctly (in which case it's weird and should be changed), or if I've implemented the proposal wrong. Or maybe the proposal is underspecified here and I've picked a weird interpretation?

Zooming in to this a bit: my updates don't touch this bit of the proposal. The proposal specifies that the BNF for types becomes `btype ::= modifiers atype | btype atype`, but doesn't say what `btype` is. Probably it's meant to compare to the [2010 report](https://www.haskell.org/onlinereport/haskell2010/haskellch10.html) which gives `type ::= btype [-> btype]` and `btype ::= [btype] atype`. So the proposal would accept `%O C a` as a `btype` (`modifiers atype atype`). But does that specify whether it would parse the same as `(%O C) a`, or as `%O (C a)`, or not specify? I'm not familiar enough with this whole area to say offhand.

(To be clear: the current implementation is clearly silly, and if it's not obvious whether it matches the proposal or not, the proposal should be updated. I'm just not immediately sure exactly how it should be updated, or how to fix the implementation.)

What would be better? We can imagine [*] would instead parse the same as

    instance (%Overlappable (C a)) => C [a] where ...

which is also currently legal. But it wouldn't parse the same as

    %Overlappable
    instance C a => C [a] where ...

and if we're implementing a substitute for `{-# OVERLAPPABLE #-}`, I'd only expect the latter to work, not [*].

We could also imagine [*] parses the same as

    instance %Overlappable (C a => C [a]) where ...

but this isn't currently accepted as an instance head (and nor would `instance %Overlappable (C a)` be).

Or we could imagine that [*] just doesn't parse, and users have to add their own parens. Or maybe we just don't support attaching modifiers to types until there's a concrete use for them. I think one of these is my favorite.

There's a similar story for exprs, where (as currently implemented) `%A id 3` parses the same as `(%A id) 3`. Maybe those shouldn't be supported either? And I'm starting to look sideways at patterns, though those might be implemented okay? (`%A C x`, `(%A C) x` and `C %A x` are forbidden; `%A (C x)` and `C (%A x)` are allowed.)

ChickenProp committed:
Remove modifiers on types, exprs and patterns.

ChickenProp committed:
Slight clarification to RN/TC modifiers.

ChickenProp committed:
-Wunrecognized-modifiers is enabled by default.

ChickenProp committed:
Allow modifiers in patterns, not just bindings.

ChickenProp committed:
';' required on decl modifiers.

ChickenProp committed:
Formatting.

ChickenProp committed:
Update incorrect wording.

I assume left over from a very early version of the proposal.

sgraf812 commented on proposals/0370-modifiers.rst:
```diff
@@ -44,136 +45,246 @@ a different syntax. Specifically, the term-level operators ``inline`` and
+
+    * The linear arrow ``a ⊸ b`` has the same meaning as ``a %1 -> b``. Other
+      modifiers are accepted: ``a %Matchable ⊸ b`` has the same meaning as ``a
+      %Matchable %1 -> b``.
```
What if I had `a %Many %One -> b`? Is this a linear or non-linear arrow?

ChickenProp replied:
With `-XLinearTypes`, that's an error. (Specified a few lines above: "Multiple modifiers of type `Multiplicity` on the same arrow are not allowed.") Without, both modifiers are ignored with a warning.

Tritlo committed:
Update proposal headers

Tritlo committed:
Merge branch 'master' into modifiers-updates
